### PR TITLE
PHP: change to awk from gawk, since mac doesn't support it

### PR DESCRIPTION
**Pull request by HannahShiSFB**
Partial revert of #23255 due to [this concern](https://github.com/grpc/grpc/pull/23255#issuecomment-662957165). [Edit:] This sort of becomes a forward fix now.

Replace `gawk` with `awk` so as not to introduce an extra dependency.

- Cannot use multidimensional array [reference](https://www.gnu.org/software/gawk/manual/html_node/Multidimensional.html)
  - Solution: use a bunch of one dimensional arrays instead
- Cannot use `gensub`
- Cannot even pass in `array` in `match(string, regexp [, array])` [reference](https://www.gnu.org/software/gawk/manual/html_node/String-Functions.html)
  - Solution: write our own `sed_gensub`

### CODE REVIEW

File: src/php/bin/php_extension_doxygen_filter.awk [hunk 1]
```diff
@@ -12,38 +12,63 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+function sed_gensub(regexp, replacement, how, target,     cmd_, ret_) {
```
Comment (stanley-cheung) [inline src/php/bin/php_extension_doxygen_filter.awk:15]: There's a bit of unnecessary space. Please remove it?

File: src/php/bin/php_extension_doxygen_filter.awk [hunk 2]
```diff
@@ -12,38 +12,63 @@
...
+    if (!target) {
+        target = $0
+    }
+    gsub(/'/, "'\"'\"'", target);
+    gsub(/\\\\/, "\\", regexp);
+
+    #cmd_ = "printf '" target "' | sed -E '/" regexp "/!{q1};{s//" replacement "/" tolower(how) "}'";
```
Comment (stanley-cheung) [inline src/php/bin/php_extension_doxygen_filter.awk:22]: Please remove this commented out line.

File: src/php/bin/php_extension_doxygen_filter.awk [hunk 3]
```diff
@@ -12,38 +12,63 @@
...
+}
+
 BEGIN {
     namespace = "Grpc";
     className = "";
     classDocComment = "";
-    delete methods; # methods[method][doc|args|static]
-    delete constants; # constants[i][name|doc]
+
+    delete methodNames; # methodNames[i]
```
Comment (stanley-cheung) [inline src/php/bin/php_extension_doxygen_filter.awk:38]: Perhaps consider commenting it this way? `# i => methodName`.

File: src/php/bin/php_extension_doxygen_filter.awk [hunk 4]
```diff
@@ -12,38 +12,63 @@
...
+    delete methodArgs; # methodArgs[methodName]
```
Comment (stanley-cheung) [inline src/php/bin/php_extension_doxygen_filter.awk:39]: Perhaps consider commenting it this way `# methodName => concatenatedArgsStr`

File: src/php/bin/php_extension_doxygen_filter.awk [hunk 5]
```diff
@@ -12,38 +12,63 @@
...
+    delete methodDocs; # methodDocs[methodName]
+    delete methodStatics; # methodStatics[methodName], 1 if static
```
Comment (stanley-cheung) [inline src/php/bin/php_extension_doxygen_filter.awk:41]: Perhaps consider commenting it this way `# methodName => 1 if static`.

File: src/php/bin/php_extension_doxygen_filter.awk [hunk 6]
```diff
@@ -12,38 +12,63 @@
...
 BEGIN {
     namespace = "Grpc";
     className = "";
     classDocComment = "";
-    delete methods; # methods[method][doc|args|static]
-    delete constants; # constants[i][name|doc]
+
+    delete methodNames; # methodNames[i]
+    delete methodArgs; # methodArgs[methodName]
+    delete methodDocs; # methodDocs[methodName]
```
Comment (stanley-cheung) [inline src/php/bin/php_extension_doxygen_filter.awk:40]: Perhaps consider commenting it this way `# methodName => methodDocCommentStr`

File: src/php/bin/php_extension_doxygen_filter.awk [hunk 7]
```diff
@@ -12,38 +12,63 @@
...
+    delete methodStatics; # methodStatics[methodName], 1 if static
+    methodsCount = 0;
+
+    delete constantNames; # constantNames[i]
+    delete constantDocs; # constantDocs[constantName]
```
Comment (stanley-cheung) [inline src/php/bin/php_extension_doxygen_filter.awk:45]: Same for these 2 lines

File: src/php/bin/php_extension_doxygen_filter.awk [hunk 8]
```diff
@@ -80,37 +104,45 @@ inDocComment==1 && /\*\// {
...
+
+    # concat args
+    if (argsCount > 0) {
+        methodArgs[method] = args[0];
+        for (i = 1; i < argsCount; i++) {
+            methodArgs[method] = methodArgs[method] ", " args[i];
+        }
     }
+
+
```
Comment (stanley-cheung) [inline src/php/bin/php_extension_doxygen_filter.awk:126]: Can remove one of these?

File: src/php/bin/php_extension_doxygen_filter.awk (hunk 1 above)
Comment (HannahShiSFB) [reply src/php/bin/php_extension_doxygen_filter.awk]: the spaces in the parameter list are necessary to separate out local variables

File: src/php/bin/php_extension_doxygen_filter.awk (hunk 2 above)
Comment (HannahShiSFB) [reply src/php/bin/php_extension_doxygen_filter.awk]: addressed

File: src/php/bin/php_extension_doxygen_filter.awk (hunk 3 above)
Comment (HannahShiSFB) [reply src/php/bin/php_extension_doxygen_filter.awk]: addressed

File: src/php/bin/php_extension_doxygen_filter.awk (hunk 4 above)
Comment (HannahShiSFB) [reply src/php/bin/php_extension_doxygen_filter.awk]: addressed

File: src/php/bin/php_extension_doxygen_filter.awk (hunk 5 above)
Comment (HannahShiSFB) [reply src/php/bin/php_extension_doxygen_filter.awk]: addressed

File: src/php/bin/php_extension_doxygen_filter.awk (hunk 6 above)
Comment (HannahShiSFB) [reply src/php/bin/php_extension_doxygen_filter.awk]: addressed

File: src/php/bin/php_extension_doxygen_filter.awk (hunk 7 above)
Comment (HannahShiSFB) [reply src/php/bin/php_extension_doxygen_filter.awk]: addressed

File: src/php/bin/php_extension_doxygen_filter.awk (hunk 8 above)
Comment (HannahShiSFB) [reply src/php/bin/php_extension_doxygen_filter.awk]: addressed

**Comment by jtattermusch** [comment]
When I ran `tools/buildgen/generate_projects.sh` on Mac, I got a different contents of the generated Doxyfile.php file

```
diff --git a/tools/doxygen/Doxyfile.php b/tools/doxygen/Doxyfile.php
index f99b55b93a..34f1476173 100644
--- a/tools/doxygen/Doxyfile.php
+++ b/tools/doxygen/Doxyfile.php
@@ -803,21 +803,21 @@ src/php/docker/README.md \
 src/php/lib/Grpc/AbstractCall.php \
 src/php/lib/Grpc/BaseStub.php \
 src/php/lib/Grpc/BidiStreamingCall.php \
-src/php/lib/Grpc/Call.php \
-src/php/lib/Grpc/CallCredentials.php \
 src/php/lib/Grpc/CallInvoker.php \
-src/php/lib/Grpc/Channel.php \
-src/php/lib/Grpc/ChannelCredentials.php \
 src/php/lib/Grpc/ClientStreamingCall.php \
 src/php/lib/Grpc/Constants.php \
 src/php/lib/Grpc/DefaultCallInvoker.php \
 src/php/lib/Grpc/Interceptor.php \
 src/php/lib/Grpc/RpcServer.php \
-src/php/lib/Grpc/Server.php \
-src/php/lib/Grpc/ServerCredentials.php \
 src/php/lib/Grpc/ServerStreamingCall.php \
-src/php/lib/Grpc/Timeval.php \
-src/php/lib/Grpc/UnaryCall.php
+src/php/lib/Grpc/UnaryCall.php \
+src/php/lib/Grpc/call.php \
+src/php/lib/Grpc/call_credentials.php \
+src/php/lib/Grpc/channel.php \
+src/php/lib/Grpc/channel_credentials.php \
+src/php/lib/Grpc/server.php \
+src/php/lib/Grpc/server_credentials.php \
+src/php/lib/Grpc/timeval.php
``` 

Not sure what's causing it, but the script must generate exactly the same output when ran on Linux or Mac, otherwise it's badly broken.

**Comment by stanley-cheung** [comment]
> When I ran `tools/buildgen/generate_projects.sh` on Mac, I got a different contents of the generated Doxyfile.php file
> 
> Not sure what's causing it, but the script must generate exactly the same output when ran on Linux or Mac, otherwise it's badly broken.

Yes I am aware of it. We are looking into this. Will need to fix this.

**Comment by stanley-cheung** [comment]
@HannahShiSFB Thanks for taking care of this additional Mac fix. I ran `generate_projects.sh` on a Mac machine again with this updated branch and there's no diff generated after this additional fix. Will let the rest of the PR tests run. Thanks!

**Pull request merged**